### PR TITLE
Add tag requirements to jobs

### PR DESCRIPTION
**Pull request by guidow**
This allows even more ways of restricting which nodes can or cannot work on
which jobs.

### CODE REVIEW

**Comment by opalmer** [comment]
Looking good so far, I assume this is still being tested though (hence the 'Experimental' bit).

**Comment by guidow** [comment]
Yes, this was an experiment at first to see if we can tackle a problem we were having with this. Turns out that works just fine, so I'm removing the [Experimental] from the title again.

File: pyfarm/master/user_interface/jobs.py [hunk 1]
```diff
@@ -937,6 +937,38 @@ def update_tags_in_job(job_id):
...
 
+def update_tag_requirements_in_job(job_id):
+    job = Job.query.filter_by(id=job_id).first()
+    if not job:
+        return (render_template(
+                    "pyfarm/error.html", error="Job %s not found" % job_id),
+                NOT_FOUND)
+
+    tagnames = request.form["tag_requirements"].split(" ")
+    tagnames = [x.strip() for x in tagnames if not x == ""]
```
Comment (opalmer) [inline pyfarm/master/user_interface/jobs.py]: Possibly cleaner?

```python
filter(bool, map(str.strip, tagnames))
```

File: pyfarm/master/user_interface/jobs.py [hunk 2]
```diff
@@ -937,6 +937,38 @@ def update_tags_in_job(job_id):
...
+    job.tag_requirements = []
+    for name in tagnames:
+        negate = False
+        if name.startswith("-"):
+            if len(name) < 2:
+                return (render_template(
+                    "pyfarm/error.html", error="Tag too short: " % name),
```
Comment (opalmer) [inline pyfarm/master/user_interface/jobs.py]: Missing a `%s` in the error message.

File: pyfarm/models/agent.py [hunk 3]
```diff
@@ -357,6 +357,26 @@ def satisfies_jobtype_requirements(self, jobtype_version):
 
         return len(requirements_to_satisfy) == 0
 
+    def satisfies_job_requirements(self, job):
+        if not self.satisfies_jobtype_requirements( job.jobtype_version):
```
Comment (opalmer) [inline pyfarm/models/agent.py]: Extra whitespace after `(`

**Comment by opalmer** [comment]
Just a couple of comments but other than that I think this is ready to merge.

**Pull request merged**